File: app/src/main/java/com/popalay/yocard/ui/home/HomeActivity.java
package com.popalay.yocard.ui.home;

import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBarDrawerToggle;
import android.view.Menu;
import android.view.MenuItem;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.popalay.yocard.Constants;
import com.popalay.yocard.R;
import com.popalay.yocard.data.events.AddCardEvent;
import com.popalay.yocard.data.events.FavoriteHolderEvent;
import com.popalay.yocard.databinding.ActivityHomeBinding;
import com.popalay.yocard.ui.base.BaseActivity;
import com.popalay.yocard.ui.cards.CardsFragment;
import com.popalay.yocard.ui.debts.DebtsFragment;
import com.popalay.yocard.ui.holders.HoldersFragment;
import com.popalay.yocard.utils.BrowserUtils;

import org.greenrobot.eventbus.EventBus;

import shortbread.Shortcut;

public class HomeActivity extends BaseActivity implements HomeView {

    private static final int MENU_SETTINGS = Menu.FIRST;

    @InjectPresenter HomePresenter presenter;

    private int startPage = R.id.cards;

    private ActivityHomeBinding b;
    private ActionBarDrawerToggle toggle;

    public static Intent getIntent(Context context) {
        return new Intent(context, HomeActivity.class);
    }

    @ProvidePresenter
    HomePresenter providePresenter() {
        return new HomePresenter(startPage);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        b = DataBindingUtil.setContentView(this, R.layout.activity_home);
        initUI();
    }

    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        toggle.syncState();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        toggle.onConfigurationChanged(newConfig);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        menu.add(Menu.NONE, MENU_SETTINGS, Menu.NONE, R.string.settings)
                .setIcon(R.drawable.ic_settings)
                .setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (toggle.onOptionsItemSelected(item)) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void openPage(int pageId) {
        final Fragment fragment;
        switch (pageId) {
            case R.id.cards:
                fragment = CardsFragment.newInstance();
                break;
            case R.id.holders:
                fragment = HoldersFragment.newInstance();
                break;
            case R.id.debts:
                fragment = DebtsFragment.newInstance();
                break;
            default:
                throw new IllegalArgumentException("Illegal position");
        }

        final Fragment current = getSupportFragmentManager().findFragmentByTag(fragment.getClass().getSimpleName());
        if (current != null) {
            getSupportFragmentManager().beginTransaction()
                    .setCustomAnimations(R.anim.slide_up, 0, R.anim.slide_up, 0)
                    .replace(R.id.host, current, current.getClass().getSimpleName())
                    .commit();
        } else {
            getSupportFragmentManager().beginTransaction()
                    .setCustomAnimations(R.anim.slide_up, 0, R.anim.slide_up, 0)
                    .add(R.id.host, fragment, fragment.getClass().getSimpleName())
                    .commit();
        }
    }

    private boolean onNavigationClick(MenuItem item) {
        presenter.onBottomNavigationItemClick(item.getItemId());
        return true;
    }

    private boolean onDrawerItemClick(MenuItem item) {
        presenter.onDrawerItemClick(item.getItemId());
        switch (item.getItemId()) {
            case R.id.navigation_privacy_policy:
                BrowserUtils.openLink(this, Constants.PRIVACY_POLICY_LINK);
                break;
        }
        b.drawerLayout.closeDrawers();
        return true;
    }

    private void initUI() {
        setSupportActionBar(b.toolbar);
        toggle = new ActionBarDrawerToggle(this, b.drawerLayout,
                R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        toggle.setDrawerIndicatorEnabled(true);
        b.drawerLayout.addDrawerListener(toggle);
        b.navigationView.setNavigationItemSelectedListener(this::onDrawerItemClick);
        b.bottomBar.setSelectedItemId(startPage);
        b.bottomBar.setOnNavigationItemSelectedListener(this::onNavigationClick);
    }

    // Shortcuts
    @Shortcut(id = "shortcut_add_card", icon = R.drawable.ic_shortcut_add_card,
            shortLabelRes = R.string.shortcut_add_card)
    public void addCardShortcut() {
        startPage = R.id.cards;
        EventBus.getDefault().postSticky(new AddCardEvent());
    }

    @Shortcut(id = "shortcut_favorite_holder", icon = R.drawable.ic_shortcut_favorite_holder,
            rank = 1, shortLabelRes = R.string.shortcut_favorite_holder)
    public void favoriteHolderShortcut() {
        startPage = R.id.holders;
        EventBus.getDefault().postSticky(new FavoriteHolderEvent());
    }

    @Shortcut(id = "shortcut_debts", icon = R.drawable.ic_shortcut_debts,
            rank = 2, shortLabelRes = R.string.shortcut_debts)
    public void debtsShortcut() {
        startPage = R.id.debts;
    }
}

File: app/src/main/kotlin/com/popalay/cardme/presentation/screens/home/HomeActivity.kt
package com.popalay.cardme.presentation.screens.home

import android.app.PendingIntent
import android.arch.lifecycle.ViewModelProvider
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.res.Configuration
import android.nfc.NdefMessage
import android.nfc.NfcAdapter
import android.nfc.tech.NfcF
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.ActionBarDrawerToggle
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import com.popalay.cardme.R
import com.popalay.cardme.domain.interactor.ShortcutInteractor
import com.popalay.cardme.databinding.ActivityHomeBinding
import com.popalay.cardme.presentation.base.BaseActivity
import com.popalay.cardme.presentation.base.navigation.CustomNavigator
import com.popalay.cardme.presentation.screens.cards.CardsFragment
import com.popalay.cardme.utils.extensions.findFragmentByType
import com.popalay.cardme.utils.extensions.getDataBinding
import com.popalay.cardme.utils.extensions.getViewModel
import com.popalay.cardme.utils.extensions.setSelectedItem
import dagger.android.DispatchingAndroidInjector
import dagger.android.support.HasSupportFragmentInjector
import shortbread.Shortcut
import javax.inject.Inject


class HomeActivity : BaseActivity(), HasSupportFragmentInjector {

    companion object {
        private val MENU_SETTINGS = Menu.FIRST
        fun getIntent(context: Context) = Intent(context, HomeActivity::class.java)
    }

    @Inject lateinit var factory: ViewModelProvider.Factory
    @Inject lateinit var androidInjector: DispatchingAndroidInjector<Fragment>
    @Inject override lateinit var navigator: CustomNavigator
    @Inject lateinit var shortcutInteractor: ShortcutInteractor
    @Inject lateinit var viewModelFacade: HomeViewModelFacade

    private lateinit var b: ActivityHomeBinding
    private lateinit var toggle: ActionBarDrawerToggle

    private var adapter: NfcAdapter? = null
    private var pendingIntent: PendingIntent? = null
    private var filters: Array<IntentFilter>? = null
    private var techLists: Array<Array<String>>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        b = getDataBinding(R.layout.activity_home)
        b.vm = getViewModel<HomeViewModel>(factory)
        initNfcListening()
        initUI()
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        toggle.syncState()
    }

    override fun onNewIntent(intent: Intent) = processIntent(intent)

    public override fun onResume() {
        super.onResume()
        adapter?.enableForegroundDispatch(this, pendingIntent, filters, techLists)
    }

    override fun onPause() {
        super.onPause()
        adapter?.disableForegroundDispatch(this)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == CardsFragment.SCAN_REQUEST_CODE) {
            findFragmentByType<CardsFragment>()?.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        toggle.onConfigurationChanged(newConfig)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menu.add(Menu.NONE, MENU_SETTINGS, Menu.NONE, R.string.settings)
                .setIcon(R.drawable.ic_settings)
                .setShowAsAction(MenuItem.SHOW_AS_ACTION_IF_ROOM)
        return true
    }

    override fun onBackPressed() {
        if (b.drawerLayout.isDrawerOpen(Gravity.START)) {
            b.drawerLayout.closeDrawers()
            return
        }
        super.onBackPressed()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (toggle.onOptionsItemSelected(item)) return true
        when (item.itemId) {
            MENU_SETTINGS -> viewModelFacade.onSettingClick()
        }
        return super.onOptionsItemSelected(item)
    }

    fun selectTab(itemId: Int) {
        b.bottomBar.setSelectedItem(itemId, false)
    }

    override fun supportFragmentInjector() = androidInjector

    private fun initUI() {
        setSupportActionBar(b.toolbar)
        toggle = ActionBarDrawerToggle(this, b.drawerLayout,
                R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        toggle.isDrawerIndicatorEnabled = true
        b.drawerLayout.addDrawerListener(toggle)
    }

    //TODO create specific class
    private fun initNfcListening() {
        adapter = NfcAdapter.getDefaultAdapter(this)
        pendingIntent = PendingIntent.getActivity(this, 0, Intent(this, javaClass)
                .addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP or Intent.FLAG_ACTIVITY_CLEAR_TOP), 0)
        val ndef = IntentFilter(NfcAdapter.ACTION_NDEF_DISCOVERED)
        try {
            ndef.addDataType("application/" + packageName)
        } catch (e: IntentFilter.MalformedMimeTypeException) {
            throw RuntimeException("fail", e)
        }
        filters = arrayOf(ndef)
        techLists = arrayOf((arrayOf<String>(NfcF::class.java.name)))
    }

    private fun processIntent(intent: Intent) {
        val rawMsgs = intent.getParcelableArrayExtra(NfcAdapter.EXTRA_NDEF_MESSAGES)
        val msg = rawMsgs [0] as NdefMessage
        val bytes = msg.records[0].payload
        viewModelFacade.onNfcMessageRead(bytes)
        getIntent().action = null
    }

    // Shortcuts
    @Shortcut(id = "SHORTCUT_ADD_CARD", icon = R.drawable.ic_shortcut_add_card, rank = 0, shortLabelRes = R.string.shortcut_add_card)
    fun addCardShortcut() {
        shortcutInteractor.applyShortcut(ShortcutInteractor.Shortcut.ADD_CARD)
    }

    @Shortcut(id = "SHORTCUT_DEBTS", icon = R.drawable.ic_shortcut_debts, rank = 2, shortLabelRes = R.string.shortcut_debts)
    fun debtsShortcut() {
        shortcutInteractor.applyShortcut(ShortcutInteractor.Shortcut.DEBTS)
    }
}
